Feature: add incremental encoder

Olavhaasie opened the pull request:
Closes PM-24

## Description
Adds the feature to read out the incremental encoder on the joints. It maps the Motor Position object from the IMotionCube.

@bjornminderman started on this feature and I solved the conflicts it had, since it was a bit outdated.

The original `Encoder` class is now a super class of `AbsoluteEncoder` and `IncrementalEncoder`. Also in order to keep it clean I have added a new `encoder` directory, where these classes reside. The `IMotionCube` and `Joint` class now have two separate functions for reading out the absolute and incremental encoder position. The incremental encoder position is only published on the imc state topic. It shouldn't be too hard to use it for any calculations in the hardware interface.

## Changes
* Create `AbsoluteEncoder` and `IncrementalEncoder`
* Rename a lot of stuff from `Encoder` to `AbsoluteEncoder`
* Add publishing incremental encoder position
* Add tests

<!-- Please don't forget to request for reviews -->

RutgerVanBeek commented on march_hardware/include/march_hardware/encoder/Encoder.h:
```diff
@@ -0,0 +1,48 @@
+  explicit Encoder(size_t number_of_bits);
+
+  /*
+   * Reads out the encoder from the slave and returns the value in Internal Units (IU).
```
Absolute or incremental encoder? or change the encoder to an encoder. Now it reads like there is only one encoder.

Olavhaasie replied:
It's both actually since this is a base class for those types of encoders. The _the_ refers to the actual encoder instance in use. So this description would be valid for both `absolute_encoder.getAngleIU()` and `incremental_encoder.getAngleIU()`

RutgerVanBeek commented on march_hardware/src/encoder/IncrementalEncoder.cpp:
```diff
@@ -0,0 +1,21 @@
+double IncrementalEncoder::toRad(int32_t iu)
+{
+  return iu * 2 * M_PI / Encoder::getTotalPositions();
+}
```
Although I am not sure how it does exactly work, I think you cannot do this as it can wrap around. I think we should leave this or a similar function out until we know what the output is. Or does @bjornminderman know more about this?

Olavhaasie replied:
I agree, I am not sure if it will be useful, for now the IU is published and the radians are not used.

Roelemans commented on march_hardware/src/encoder/IncrementalEncoder.cpp:
```diff
@@ -0,0 +1,21 @@
+double IncrementalEncoder::getAngleRad(uint8_t byte_offset)
+{
+  return this->toRad(Encoder::getAngleIU(byte_offset));
+}
```
You can just move this to the superclass since AbsoluteEncoder contains an identical method

Roelemans commented on march_hardware/src/encoder/IncrementalEncoder.cpp:
```diff
@@ -0,0 +1,21 @@
+double IncrementalEncoder::toRad(int32_t iu)
+{
+  return iu * 2 * M_PI / Encoder::getTotalPositions();
+}
```
I think you can do the same here if you set zero_position_iu to zero.

Olavhaasie commented:
I will study the behavior of the incremental encoders first

Olavhaasie commented:
I have added a `transmission` variable to the incremental encoder config in https://github.com/project-march/hardware-interface/pull/225/commits/b53791052fb7ff869baa17dbf90d4f691b60dc13. The transmission is already defined in the xacro, however, I was not able to access that with the `urdf` package, so I added it as a variable in the yaml.

Roelemans commented on march_hardware/src/encoder/IncrementalEncoder.cpp:
```diff
@@ -16,6 +17,6 @@ double IncrementalEncoder::getAngleRad(uint8_t byte_offset)
 
```
Move above method to super class? :innocent:

Roelemans commented on march_hardware/src/encoder/IncrementalEncoder.cpp:
```diff
@@ -16,6 +17,6 @@ double IncrementalEncoder::getAngleRad(uint8_t byte_offset)
 double IncrementalEncoder::toRad(int32_t iu)
 {
-  return iu * 2 * M_PI / Encoder::getTotalPositions();
+  return iu * this->transmission_ * 2 * M_PI / Encoder::getTotalPositions();
```
```suggestion
  return iu * this->transmission_ * PI_2 / Encoder::getTotalPositions();
```

Roelemans commented on march_hardware/test/encoder/TestIncrementalEncoder.cpp:
```diff
@@ -20,6 +21,6 @@ TEST_F(TestIncrementalEncoder, ZeroIUToRad)
 {
   const int32_t iu = 1000;
-  const double expected = iu * 2.0 * M_PI / std::pow(2, this->resolution);
+  const double expected = iu * this->transmission * 2.0 * M_PI / std::pow(2, this->resolution);
```
```suggestion
  const double expected = iu * this->transmission * PI_2 / std::pow(2, this->resolution);
```
